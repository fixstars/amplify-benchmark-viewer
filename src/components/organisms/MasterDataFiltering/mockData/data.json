[
  { "name": "Tsp" },
  { "name": "Qap" },
  { "name": "MaxCut" },
  { "name": "Sudoku" },
  { "name": "Cvrp" }
]
