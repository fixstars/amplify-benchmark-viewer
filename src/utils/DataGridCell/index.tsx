/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { LinkButton } from 'components/atoms/LinkButton'

const clientColors = [
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
]

const instanceColors = [
  '#ff66ff',
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#718de6',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
  '#9a916d',
  '#406a0f',
  '#004b31',
  '#18729f',
  '#9997c3',
  '#9cf0ec',
  '#ce843c',
  '#2d326e',
  '#76ed60',
  '#d7f7e8',
  '#bb15d7',
  '#87a7d8',
  '#94053a',
  '#e5c8a6',
  '#e2f2c0',
]

export interface ProblemInstances {
  readonly [key: string]: ReadonlyArray<string> | undefined
}

const renderClientsCell = (key: string, clients: ReadonlyArray<string>) => {
  const getBackgroundColor = (client: string): string => {
    const cache = JSON.parse(localStorage.getItem('clientColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('clientColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[client]

    if (backgroundColor === undefined) {
      const length = clientColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = clientColors[cacheIndex]
      cache[client] = backgroundColor
      localStorage.setItem('clientColorCache', JSON.stringify(cache))
      localStorage.setItem('clientColorIndexCache', JSON.stringify(cacheIndex))
      return backgroundColor
    }

    return backgroundColor
  }

  return clients.map((client) => (
    <div key={`${key}_${client}`} style={{ marginRight: 10 }}>
      <LinkButton
        label={client}
        link={`/clients/${client}`}
        backgroundColor={getBackgroundColor(client)}
      />
    </div>
  ))
}

const renderProblemInstancesCell = (
  key: string,
  problemInstances: ProblemInstances,
) => {
  const getBackgroundColor = (instance: string): string => {
    const cache = JSON.parse(localStorage.getItem('instanceColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('instanceColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[instance]

    if (backgroundColor === undefined) {
      const length = instanceColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = instanceColors[cacheIndex]
      cache[instance] = backgroundColor
      localStorage.setItem('instanceColorCache', JSON.stringify(cache))
      localStorage.setItem(
        'instanceColorIndexCache',
        JSON.stringify(cacheIndex),
      )
      return backgroundColor
    }

    return backgroundColor
  }
  const pairs = Object.entries(problemInstances)
  pairs.sort(function (p1, p2) {
    const p1Key = p1[0]
    const p2Key = p2[0]
    if (p1Key < p2Key) {
      return -1
    }
    if (p1Key > p2Key) {
      return 1
    }
    return 0
  })
  problemInstances = Object.fromEntries(pairs)
  return Object.keys(problemInstances).map((className) => (
    <div
      key={`${key}_${className}`}
      style={{ margin: '5px 0 5px 0', display: 'flex' }}
    >
      <span style={{ marginRight: 10 }}>{className}:</span>
      <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        {problemInstances[className]?.map((instance) => (
          <div
            key={`${key}_${className}_${instance}`}
            style={{ marginRight: 10, marginBottom: 10 }}
          >
            <LinkButton
              label={instance}
              link={`/classes/${className}/instances/${instance}`}
              backgroundColor={getBackgroundColor(instance)}
            />
          </div>
        ))}
      </div>
    </div>
  ))
}

export { renderClientsCell, renderProblemInstancesCell }
