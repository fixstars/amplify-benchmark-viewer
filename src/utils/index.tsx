export * from './DataGridCell'
export * from './statistics'
