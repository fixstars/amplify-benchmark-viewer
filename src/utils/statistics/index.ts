function quartile(data: readonly number[], q: number) {
  const sortedData = data.slice().sort((a, b) => a - b)
  const pos = Math.floor((sortedData.length - 1) * q)
  return sortedData[pos]
}
function quartile0(data: readonly number[]) {
  return quartile(data, 0)
}
function quartile25(data: readonly number[]) {
  return quartile(data, 0.25)
}

function quartile50(data: readonly number[]) {
  return quartile(data, 0.5)
}

function quartile75(data: readonly number[]) {
  return quartile(data, 0.75)
}

function quartile100(data: readonly number[]) {
  return quartile(data, 1)
}

export { quartile0, quartile25, quartile50, quartile75, quartile100 }
