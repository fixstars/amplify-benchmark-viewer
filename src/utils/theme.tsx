/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { createTheme } from '@mui/material/styles'

export const theme = createTheme({
  palette: {
    primary: {
      main: '#111',
    },
  },
  typography: {
    h2: {
      fontSize: '1.5rem',
    },
  },
  components: {
    MuiTooltip: {
      styleOverrides: {
        tooltip: {
          backgroundColor: '#111',
          color: '#f4f1f4',
          font: '10px',
        },
      },
    },
  },
})
