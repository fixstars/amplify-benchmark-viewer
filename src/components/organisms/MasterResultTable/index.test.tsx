import { render, screen, fireEvent, waitFor } from '@testing-library/react'
import type { MasterData } from '@types'
import { mockDataGrid, mockTooltip } from 'utils/test'
import mockData from 'utils/test/mocks/data/mockMasterData.json'

import { MasterResultTable } from '.'

describe('<MasterResultTable />', () => {
  it('Rendered well', async () => {
    const { container } = render(
      <MasterResultTable data={mockData as ReadonlyArray<MasterData>} />,
    )

    const dataGrid = mockDataGrid.mock.calls[0][0]
    expect(dataGrid.sx).toEqual({
      '&.MuiDataGrid-root .MuiDataGrid-cell:focus-within': {
        outline: 'none !important',
      },
    })

    expect(dataGrid.rows).toEqual([
      {
        id: '0-hhhhhhhhhhhhhhhhhhhhhhhhhh',
        num_feasibles: 20,
        mean: 100455.8,
        std: 1194.797476782704,
        min: 98340,
        '25%': 99680.25,
        '50%': 100536,
        '75%': 101316,
        max: 102254,
        num_samples: 20,
        num_reach_best: 0,
        setting_time: 1000000,
      },
      {
        id: '0-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
        num_feasibles: 20,
        mean: 61958.5,
        std: 1127.9892916624976,
        min: 60150,
        '25%': 61323.75,
        '50%': 61608,
        '75%': 62679.25,
        max: 64612,
        num_samples: 20,
        num_reach_best: 0,
        setting_time: 1000000,
      },
      {
        id: '0-eeeeeeeeeeeeeeeeeeeeeeeeee',
        num_feasibles: 20,
        mean: 100455.8,
        std: 1194.797476782704,
        min: 98340,
        '25%': 99680.25,
        '50%': 100536,
        '75%': 101316,
        max: 102254,
        num_samples: 20,
        num_reach_best: 0,
        setting_time: 1000000,
      },
    ])

    expect(dataGrid.columnVisibilityModel).toEqual({
      problem_class: true,
      problem_instance: true,
    })
    expect(dataGrid.columns.length).toBe(13)

    const column1 = dataGrid.columns[0]
    expect(column1.field).toBe('problem_class')
    expect(column1.headerName).toBe('Problem class')
    expect(column1.sortable).toBe(false)
    expect(column1.minWidth).toBe(120)
    expect(column1.renderCell.name).toBe('renderCell')

    const column2 = dataGrid.columns[1]
    expect(column2.field).toBe('problem_instance')
    expect(column2.headerName).toBe('Problem instance')
    expect(column2.sortable).toBe(false)
    expect(column2.minWidth).toBe(140)
    expect(column2.renderCell.name).toBe('renderCell')

    const column3 = dataGrid.columns[2]
    expect(column3.field).toBe('client_name')
    expect(column3.headerName).toBe('Client name')
    expect(column3.sortable).toBe(false)
    expect(column3.minWidth).toBe(120)
    expect(column3.renderCell.name).toBe('renderCell')

    const column4 = dataGrid.columns[3]
    expect(column4.field).toBe('client_version')
    expect(column4.headerName).toBe('Client version')
    expect(column4.sortable).toBe(false)
    expect(column4.minWidth).toBe(200)
    expect(column4.renderCell.name).toBe('renderCell')

    const column5 = dataGrid.columns[4]
    expect(column5.field).toBe('setting_time')
    expect(column5.headerName).toBe('Setting time')
    expect(column5.sortable).toBe(false)
    expect(column5.minWidth).toBe(120)
    expect(column5.renderCell.name).toBe('renderCell')

    const column6 = dataGrid.columns[5]
    expect(column6.field).toBe('num_samples')
    expect(column6.headerName).toBe('Num samples')
    expect(column6.sortable).toBe(false)
    expect(column6.minWidth).toBe(120)
    expect(column6.renderCell.name).toBe('renderCell')

    const column7 = dataGrid.columns[6]
    expect(column7.field).toBe('num_feasibles')
    expect(column7.headerName).toBe('Num feasibles')
    expect(column7.sortable).toBe(false)
    expect(column7.minWidth).toBe(120)
    expect(column7.renderCell.name).toBe('renderCell')

    const column8 = dataGrid.columns[7]
    expect(column8.field).toBe('num_reach_best')
    expect(column8.headerName).toBe('Num reach best')
    expect(column8.sortable).toBe(false)
    expect(column8.minWidth).toBe(140)
    expect(column8.renderCell.name).toBe('renderCell')

    const column9 = dataGrid.columns[8]
    expect(column9.field).toBe('min')
    expect(column9.headerName).toBe('Min')
    expect(column9.sortable).toBe(false)
    expect(column9.minWidth).toBe(120)
    expect(column9.renderCell.name).toBe('renderCell')

    const column10 = dataGrid.columns[9]
    expect(column10.field).toBe('25%')
    expect(column10.headerName).toBe('25%')
    expect(column10.sortable).toBe(false)
    expect(column10.minWidth).toBe(120)
    expect(column10.renderCell.name).toBe('renderCell')

    const column11 = dataGrid.columns[10]
    expect(column11.field).toBe('50%')
    expect(column11.headerName).toBe('50%')
    expect(column11.sortable).toBe(false)
    expect(column11.minWidth).toBe(120)
    expect(column11.renderCell.name).toBe('renderCell')

    const column12 = dataGrid.columns[11]
    expect(column12.field).toBe('75%')
    expect(column12.headerName).toBe('75%')
    expect(column12.sortable).toBe(false)
    expect(column12.minWidth).toBe(120)
    expect(column12.renderCell.name).toBe('renderCell')

    const column13 = dataGrid.columns[12]
    expect(column13.field).toBe('max')
    expect(column13.headerName).toBe('Max')
    expect(column13.sortable).toBe(false)
    expect(column13.minWidth).toBe(120)
    expect(column13.renderCell.name).toBe('renderCell')

    expect(dataGrid.initialState).toEqual({ pagination: { pageSize: 3 } })
    expect(dataGrid.getRowId.name).toBe('getRowId')
    expect(dataGrid.hideFooter).toBe(true)
    expect(dataGrid.disableColumnMenu).toBe(true)
    expect(dataGrid.experimentalFeatures).toEqual({ columnGrouping: true })
    expect(dataGrid.columnGroupingModel).toEqual([
      {
        groupId: 'Basic information',
        children: [
          { field: 'problem_class' },
          { field: 'problem_instance' },
          { field: 'client_name' },
          { field: 'client_version' },
          { field: 'setting_time' },
          { field: 'num_samples' },
          { field: 'num_feasibles' },
          { field: 'num_reach_best' },
        ],
      },
      {
        groupId: 'Target energy',
        children: [
          { field: 'min' },
          { field: '25%' },
          { field: '50%' },
          { field: '75%' },
          { field: 'max' },
        ],
      },
    ])

    expect(container).toMatchSnapshot()
  })

  it('Show parameters tooltip well', async () => {
    render(<MasterResultTable data={mockData as ReadonlyArray<MasterData>} />)

    mockTooltip.mockClear()
    fireEvent.mouseEnter(screen.getByText('AClient'))

    await waitFor(() => {
      const tooltip = screen.getByRole('tooltip')
      expect(tooltip).toBeVisible()

      const tooltipContents = tooltip.textContent
      expect(tooltipContents?.includes('Client parameters')).toBe(true)
      expect(tooltipContents?.includes('- outputs.feasibilities: true')).toBe(
        true,
      )
      expect(tooltipContents?.includes('Problem parameters')).toBe(true)
      expect(tooltipContents?.includes('- constraint_weight: 1')).toBe(true)
      expect(tooltipContents?.includes('- seed: null')).toBe(true)
    })
  })

  it('Hide problem columns', async () => {
    render(
      <MasterResultTable
        hideProblem
        data={mockData as ReadonlyArray<MasterData>}
      />,
    )

    const dataGrid = mockDataGrid.mock.calls[0][0]
    expect(dataGrid.columnVisibilityModel).toEqual({
      problem_class: false,
      problem_instance: false,
    })
  })
})
