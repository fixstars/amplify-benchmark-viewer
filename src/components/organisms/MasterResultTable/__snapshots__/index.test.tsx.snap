// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MasterResultTable /> Rendered well 1`] = `
<div>
  <div
    aria-colcount="13"
    aria-multiselectable="false"
    aria-rowcount="5"
    class="MuiDataGrid-root MuiDataGrid-root--densityStandard css-15i6hkc-MuiDataGrid-root"
    role="grid"
  >
    <div>
      <div />
    </div>
    <div
      class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
    >
      <div
        class="MuiDataGrid-columnHeaders css-f3jnds-MuiDataGrid-columnHeaders"
        style="min-height: 112px; max-height: 112px; line-height: 56px;"
      >
        <div
          class="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-gl260s-MuiDataGrid-columnHeadersInner"
          role="rowgroup"
          style="transform: translate3d(0px, 0px, 0px);"
        >
          <div
            aria-rowindex="1"
            class="css-yrdy0g-MuiDataGrid-columnHeaderRow"
            role="row"
            style="height: 56px; transform: translateX(-0px);"
          >
            <div
              aria-colindex="1"
              aria-colspan="8"
              aria-label="Basic information"
              aria-sort="none"
              class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--filledGroup"
              data-fields="|-problem_class-|-problem_instance-|-client_name-|-client_version-|-setting_time-|-num_samples-|-num_feasibles-|-num_reach_best-|"
              role="columnheader"
              style="height: 56px; width: 1080px; min-width: 1080px; max-width: 1080px;"
              tabindex="-1"
            >
              <div
                class="MuiDataGrid-columnHeaderDraggableContainer"
                draggable="false"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainer"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainerContent"
                  >
                    <div
                      aria-label=""
                      class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                      data-mui-internal-clone-element="true"
                    >
                      Basic information
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-rowindex="2"
            class="css-yrdy0g-MuiDataGrid-columnHeaderRow"
            role="row"
          >
            <div
              aria-colindex="1"
              aria-label="Problem class"
              aria-sort="none"
              class="MuiDataGrid-columnHeader"
              data-field="problem_class"
              role="columnheader"
              style="height: 56px; width: 120px; min-width: 120px; max-width: 120px;"
              tabindex="0"
            >
              <div
                class="MuiDataGrid-columnHeaderDraggableContainer"
                draggable="false"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainer"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainerContent"
                  >
                    <div
                      aria-label=""
                      class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                      data-mui-internal-clone-element="true"
                    >
                      Problem class
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                style="min-height: 56px; opacity: 1;"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                  data-testid="SeparatorIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 19V5h2v14z"
                  />
                </svg>
              </div>
            </div>
            <div
              aria-colindex="2"
              aria-label="Problem instance"
              aria-sort="none"
              class="MuiDataGrid-columnHeader"
              data-field="problem_instance"
              role="columnheader"
              style="height: 56px; width: 140px; min-width: 140px; max-width: 140px;"
              tabindex="-1"
            >
              <div
                class="MuiDataGrid-columnHeaderDraggableContainer"
                draggable="false"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainer"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainerContent"
                  >
                    <div
                      aria-label=""
                      class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                      data-mui-internal-clone-element="true"
                    >
                      Problem instance
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                style="min-height: 56px; opacity: 1;"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                  data-testid="SeparatorIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 19V5h2v14z"
                  />
                </svg>
              </div>
            </div>
            <div
              aria-colindex="3"
              aria-label="Client name"
              aria-sort="none"
              class="MuiDataGrid-columnHeader"
              data-field="client_name"
              role="columnheader"
              style="height: 56px; width: 120px; min-width: 120px; max-width: 120px;"
              tabindex="-1"
            >
              <div
                class="MuiDataGrid-columnHeaderDraggableContainer"
                draggable="false"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainer"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainerContent"
                  >
                    <div
                      aria-label=""
                      class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                      data-mui-internal-clone-element="true"
                    >
                      Client name
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                style="min-height: 56px; opacity: 1;"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                  data-testid="SeparatorIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 19V5h2v14z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        style="overflow: visible; height: 0px; width: 0px;"
      >
        <div
          class="MuiDataGrid-virtualScroller css-1w5m2wr-MuiDataGrid-virtualScroller"
          style="width: 0px; height: auto; margin-top: 112px; overflow-x: hidden;"
        >
          <div
            class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
            style="width: auto; height: 156px; min-height: auto;"
          >
            <div
              class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
              style="transform: translate3d(0px, 0px, 0px);"
            >
              <div
                aria-rowindex="3"
                aria-selected="false"
                class="MuiDataGrid-row"
                data-id="0-hhhhhhhhhhhhhhhhhhhhhhhhhh"
                data-rowindex="0"
                role="row"
                style="max-height: 52px; min-height: 52px;"
              >
                <div
                  aria-colindex="1"
                  aria-colspan="1"
                  class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                  data-colindex="0"
                  data-field="problem_class"
                  role="cell"
                  style="min-width: 120px; max-width: 120px; min-height: 52px; max-height: 52px;"
                  tabindex="-1"
                >
                  <span
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    abc
                  </span>
                </div>
                <div
                  aria-colindex="2"
                  aria-colspan="1"
                  class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                  data-colindex="1"
                  data-field="problem_instance"
                  role="cell"
                  style="min-width: 140px; max-width: 140px; min-height: 52px; max-height: 52px;"
                  tabindex="-1"
                >
                  <span
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    b_instance
                  </span>
                </div>
                <div
                  aria-colindex="3"
                  aria-colspan="1"
                  class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                  data-colindex="2"
                  data-field="client_name"
                  role="cell"
                  style="min-width: 120px; max-width: 120px; min-height: 52px; max-height: 52px;"
                  tabindex="-1"
                >
                  <span
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    AClient
                  </span>
                </div>
              </div>
              <div
                aria-rowindex="4"
                aria-selected="false"
                class="MuiDataGrid-row"
                data-id="0-eeeeeeeeeeeeeeeeeeeeeeeeee"
                data-rowindex="1"
                role="row"
                style="max-height: 52px; min-height: 52px;"
              >
                <div
                  aria-colindex="1"
                  aria-colspan="1"
                  class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                  data-colindex="0"
                  data-field="problem_class"
                  role="cell"
                  style="min-width: 120px; max-width: 120px; min-height: 52px; max-height: 52px;"
                  tabindex="-1"
                >
                  <span
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    Tsp
                  </span>
                </div>
                <div
                  aria-colindex="2"
                  aria-colspan="1"
                  class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                  data-colindex="1"
                  data-field="problem_instance"
                  role="cell"
                  style="min-width: 140px; max-width: 140px; min-height: 52px; max-height: 52px;"
                  tabindex="-1"
                >
                  <span
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    b_instance
                  </span>
                </div>
                <div
                  aria-colindex="3"
                  aria-colspan="1"
                  class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                  data-colindex="2"
                  data-field="client_name"
                  role="cell"
                  style="min-width: 120px; max-width: 120px; min-height: 52px; max-height: 52px;"
                  tabindex="-1"
                >
                  <span
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    a_client
                  </span>
                </div>
              </div>
              <div
                aria-rowindex="5"
                aria-selected="false"
                class="MuiDataGrid-row MuiDataGrid-row--lastVisible"
                data-id="0-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
                data-rowindex="2"
                role="row"
                style="max-height: 52px; min-height: 52px;"
              >
                <div
                  aria-colindex="1"
                  aria-colspan="1"
                  class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                  data-colindex="0"
                  data-field="problem_class"
                  role="cell"
                  style="min-width: 120px; max-width: 120px; min-height: 52px; max-height: 52px;"
                  tabindex="-1"
                >
                  <span
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    Tsp
                  </span>
                </div>
                <div
                  aria-colindex="2"
                  aria-colspan="1"
                  class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                  data-colindex="1"
                  data-field="problem_instance"
                  role="cell"
                  style="min-width: 140px; max-width: 140px; min-height: 52px; max-height: 52px;"
                  tabindex="-1"
                >
                  <span
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    r_instance
                  </span>
                </div>
                <div
                  aria-colindex="3"
                  aria-colspan="1"
                  class="MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                  data-colindex="2"
                  data-field="client_name"
                  role="cell"
                  style="min-width: 120px; max-width: 120px; min-height: 52px; max-height: 52px;"
                  tabindex="-1"
                >
                  <span
                    class=""
                    data-mui-internal-clone-element="true"
                  >
                    a_client
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="Mui-resizeTriggers"
      >
        <div
          class="expand-trigger"
        >
          <div
            style="width: 1px; height: 1px;"
          />
        </div>
        <div
          class="contract-trigger"
        />
      </div>
    </div>
  </div>
</div>
`;
